### PR TITLE
chapter12

### PR DESCRIPTION
**Pull request by meimei0111**
ready

### CODE REVIEW

**Comment by meimei0111** [comment]
ready

File: app/mailers/user_mailer.rb [hunk 1]
```diff
@@ -4,8 +4,8 @@ def account_activation user
     mail to: user.email, subject: t(".acct_active")
   end
 
-  def password_reset
-    @greeting = t ".greeting"
-    mail to: "to@example.org"
+	def password_reset user
+		@user = user
```
Comment (ngocvt-0484) [inline app/mailers/user_mailer.rb:8]: tab 2 spaces please

Comment (meimei0111) [reply app/mailers/user_mailer.rb:8]: done

**Comment by meimei0111** [comment]
ready

File: app/mailers/user_mailer.rb [hunk 2]
```diff
@@ -1,11 +1,11 @@
-class UserMailer < ApplicationMailer
+lass UserMailer < ApplicationMailer
```
Comment (manhnt-0960) [inline app/mailers/user_mailer.rb]: error typing, Class

File: app/mailers/user_mailer.rb (hunk 1 above)
Comment (manhnt-0960) [reply app/mailers/user_mailer.rb:8]: oh, it's look not done

**Comment by meimei0111** [comment]
Screenshot from 2020-10-04 23-28-33

**Comment by meimei0111** [comment]
ready

File: app/mailers/user_mailer.rb [hunk 3]
```diff
@@ -1,11 +1,11 @@
-class UserMailer < ApplicationMailer
+Class UserMailer < ApplicationMailer
```
Comment (ngocvt-0484) [inline app/mailers/user_mailer.rb]:
```suggestion
class UserMailer < ApplicationMailer
```

**Comment by meimei0111** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+class PasswordResetsController < ApplicationController
+  before_action :load_user, only: %i(create edit update)
```
Comment (vanvtt-0952) [inline app/controllers/password_resets_controller.rb]: ` %i(create edit update)` change to ` %i(edit update)`

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def load_user
+    @user = User.find_by email: params[:email]
+  end
+
+  def valid_user
+    return if @user && @user.activated? && @user.authenticated?(:reset, params[:id])
```
Comment (vanvtt-0952) [inline app/controllers/password_resets_controller.rb]: remove ` @user &&` code

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def load_user
+    @user = User.find_by email: params[:email]
```
Comment (vanvtt-0952) [inline app/controllers/password_resets_controller.rb:44]: add code check `user is nil` by
`return if @user`
`flash[:warning] = "user not found"`
`redirect_to root_path`

**Comment by meimei0111** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**